growattServer/__init__.py

```diff
@@ -128,106 +132,439 @@ def login(self, username, password, is_password_hashed=False):
         if not is_password_hashed:
             password = hash_password(password)
 
-        response = self.session.post(self.get_url('newTwoLoginAPI.do'), data={
+        response = self.session.post(self.__get_url('newTwoLoginAPI.do'), data={
             'userName': username,
             'password': password
         })
-        data = json.loads(response.content.decode('utf-8'))['back']
-        if data['success']:
-            data.update({
-                'userId': data['user']['id'],
-                'userLevel': data['user']['rightlevel']
-            })
-        return data
+
+        if response.status_code == 200:
+            data = response.json()['back']
+            if data['success']:
+                data.update({
+                    'userId': data['user']['id'],
+                    'userLevel': data['user']['rightlevel']
+                })
+            return data
+        else:
+            raise Exception('Login failed')
 
     def plant_list(self, user_id):
         """
         Get a list of plants connected to this account.
-        """
-        response = self.session.get(self.get_url('PlantListAPI.do'),
-                                    params={'userId': user_id},
-                                    allow_redirects=False)
 
-        data = json.loads(response.content.decode('utf-8'))
-        return data['back']
+        Args:
+            user_id (str): The ID of the user.
+
+        Returns:
+            list: A list of plants connected to the account.
+
+        Raises:
+            Exception: If the request to the server fails.
+        """
+        response = self.session.get(
+            self.__get_url('PlantListAPI.do'),
+            params={'userId': user_id},
+            allow_redirects=False
+        )
+
+        if response.status_code == 200:
+            data = response.json()
+            return data.get('back', [])
+        else:
+            raise Exception('Failed to retrieve plant list')
 
     def plant_detail(self, plant_id, timespan, date=None):
         """
         Get plant details for specified timespan.
+
+        Args:
+            plant_id (str): The ID of the plant.
+            timespan (Timespan): The ENUM value conforming to the time window you want e.g. hours from today, days, or months.
+            date (datetime, optional): The date you are interested in. Defaults to datetime.datetime.now().
+
+        Returns:
+            dict: A dictionary containing the plant details.
+
+        Raises:
+            ValueError: If the timespan is not a valid Timespan.
+            Exception: If the request to the server fails.
         """
+        if not isinstance(timespan, Timespan):
+            raise ValueError("Invalid timespan value")
```

Comment:
We already check wether the timespan is a valid value in
https://github.com/indykoning/PyPi_GrowattServer/blob/75b6b73d3aa7f4014ce6759a3067d5c1375e74f3/growattServer/__init__.py#L49C10-L49C16

Though that piece of code do good to throw a message as well like you do here instead of just throwing an error

Reply:
Yeah i think that's the best approach to keep consistency 🙂